fpdf/html.py

```diff
@@ -477,7 +504,7 @@ def _end_paragraph(self):
 
     def _write_paragraph(self, text, link=None):
         if not self._paragraph:
-            self._new_paragraph()
+            self._new_paragraph(top_margin=self.font_size / self.pdf.k)
```

Comment:
Sorry I do not see where in `HTML2FPDF._new_paragraph()` this is done...

It used to be the case before this PR, yes.
But as part of this PR I removed this default value, prefering an explicit value when `_new_paragraph()` is called there, but otherwise allow a value of zero to be provided as `top_margin`.
Also I wanted to get rid of `self.follows_heading`.

Reply:
I don't think you should have removed [this](https://github.com/py-pdf/fpdf2/pull/1217/files#diff-4598b3d5f07c3ec0d147ea07e83137a1d6dd099f28a26f7c313f30708811a2a0L452-L453).

`self.follows_heading` is necessary, because just like at the top of the page, the empty line between paragraphs should not be added after a heading. The text regions handle the top-of-the-page case, but here we're dealing with a content specific issue that we need to solve where we have the necessary data available.
HTML paragraphs are more tricky than they look. They are not defined with fixed margins, but with a _distance between_ them. The existing solution implemented the standard HTML behaviour quite closely. What do we gain by removing that?

Reply:
> What do we gain by removing that?

We gain a lot of simplicity in our code, and in the expected behaviour for end-users!
That's really not something to underestimate, IMHO.

Though, I agree that we loose a bit of backward-compatibility there.

And regarding the match with the HTML standard, I made this small test:
```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Testing margins with headings - fpdf2 PR 1217</title>
    <style>
    h1, h2 { margin:  1rem 0; border: 1px solid red; }
    p      { margin: .5rem 0; border: 1px solid blue; }
    </style>
  </head>
  <body>
  <h1>Top-heading</h1>
  <p>Top-heading content paragraph 1</p>
  <p>Top-heading content paragraph 2</p>
    <h2>Sub-heading</h2>
    <p>Sub-heading content paragraph 1</p>
    <p>Sub-heading content paragraph 2</p>
  </body>
</html>
```
When rendering this HTML snippet in a browser (Firefox),
it actually seems to me that **all element margins are merged**,
without any difference based on the HTML tag type (`<hX>` or `<p>`).

To be more explicit:
* when a `<hX>` element with margin `1rem` gets rendered above a `<p>` element with margin `.5rem`, there is a vertical space of height `1rem` between them.
* when a `<p>` element with margin `.5rem` gets rendered above a `<p>` element with margin `.5rem`, there is a vertical space of height `.5rem` between them.

I checked the official specifications regarding this: the [**Margins section** of the _CSS Box Model Module Level 3_ spec](https://www.w3.org/TR/css-box-3/#margins) refers to the [**Collapsing margins** section of the CSS2 spec](https://www.w3.org/TR/CSS2/box.html#collapsing-margins), that states:

> In CSS, the adjoining margins of two or more boxes (which might or might not be siblings) can combine to form a single margin. Margins that combine this way are said to **collapse**, and the resulting combined margin is called a **collapsed margin**.

My conclusion, based on this small experiment, is that **neither the current approach on `master` nor my approach in this MR** conforms to the HTML standard.

I agree that we should aim to honor the HTML spec whenever possible.
In the case of this MR, IMHO it is better to adopt simpler code for now,
and implement "margin collapsing" later on.

What do you think of this @gmischler?

Reply:
So the result is:

* Backwards incompatibility in the output formatting (which will likely be reversed again later).
* Reduced HTML standards conformity. While the current solution is less conformant than collapsing margins everywhere, it is still more conformant than collapsing nowhere.
* A _**temporary**_ simplification of the code.

While this feature is obviously incomplete, that is not a reason to remove it, but to (eventually) expand on it to create a more complete solution. That solution will use an almost identical mechanism, not specific to headings, but remembering the most recent bottom margin, which `_new_paragraph()` will then use to reduce the provided top margin by however much they overlap. Removing the current code will make it necessary to re-invent essentially the same thing later, while giving our users a very bumpy ride of _**repeated backwards incompability**_ across all those incoherent changes.

I still don't see the benefit of that.

Reply:
Again, I'm not convinced by your arguments, but I'm getting tired of this controversy, and I'm glad to give up:
I'm going to restore `.follows_heading`, and its related code.

I'm just a bit sad for `fpdf2` end-users that will scratch their head, wondering why they have no control over the top margin of paragraphs after headings, even when they explicitly provide a `top_margin` through `tag_styles`...